PulseAudio Support

SWW13 opened the pull request:
I started working on PulseAudio support (#20).

Currently the connection to the PulseAudio server works. Retrieving the current state will require callbacks as the PulseAudio library doesn't offer synchronous calls. Besides getting the callbacks to work the whole sound block should be refactored to work with different audio systems.

Current State:
- [x] retrieve volume / mute state
- [x] listen to state changes
- [x] increase / decrease volume
- [x] mute / unmute
- [x] pulse audio / alsa fallback
- [x] clean up
- [x] documentation

atheriel commented:
The general approach looks fine to me, though we might want to make PulseAudio support opt-in. What's the advantage of routing through PulseAudio instead of using ALSA directly?

atheriel commented on src/blocks/sound.rs:
```diff
@@ -177,48 +389,28 @@ impl ConfigBlock for Sound {
-            devices: vec![SoundDevice::new("Master")?],
+            devices: vec![
+                // TODO: find better solution for mixed types
+                match PulseAudioSoundDevice::new(0) {
```
It seems unlikely that users will have mixed types if they are opting in to PulseAudio for one. However, we clearly need a way of configuring the device in use here. I don't think we should be controlling the volume of multiple devices from the same block.

SWW13 commented:
> The general approach looks fine to me, though we might want to make PulseAudio support opt-in.

Then I'll add a feature-gate for PulseAudio.

> What's the advantage of routing through PulseAudio instead of using ALSA directly?

Many advanced features such as per application audio settings (see https://en.wikipedia.org/wiki/PulseAudio#Features).

SWW13 commented:
I'm about to finish the PulseAudio support, could you please review the current state?
Some notes about the changes: 
- I also cleaned up the ALSA code and added device name support for both device types.
- I couldn't test the ALSA events as they are not working on my system because of the PulseAudio usage.
- The PulseAudio Device uses a global state (with lazy_static) because `libpulse_binding` needs static-lifetime callbacks and does not offer a synchronous API.
- There are two threads running with a PulseAudio server connection, one for sending requests and one for running the event loop to receive callbacks for subscribed events. There are two needed to block for new requests and for new events at the same time. That's the cleanest solution I came up with so far without the need to pull for events.

I have been using the PulseAudio implementation now for some days without any problems.

1wilkens commented:
FWIW, I also use this branch for a few days and it greatly improves the usefulness of the bar in pulseaudio setups, so I would like to see this merged.
Is anything else missing besides documentation?

atheriel commented:
I can confirm that this seems to be working, and that the compile-time feature flag seems to work as well. I'm a little uncomfortable with the complexity of the code, and I don't know if I have the bandwidth to really understand it. However, I don't think that's necessarily a barrier to merging it.

The one change that I would like to see is a move away from the PulseAudio -> ALSA fallback system. I'd prefer that users be able to chose their "driver" via the configuration. Could we have an explicit `driver = "pulseaudio" | "alsa" | ""` configuration option?

SWW13 commented:
> I'm a little uncomfortable with the complexity of the code, and I don't know if I have the bandwidth to really understand it. However, I don't think that's necessarily a barrier to merging it.

I'm also not happy with the complexity, but I don't see how it could be reduced much further. 

> Could we have an explicit `driver = "pulseaudio" | "alsa" | ""` configuration option?

That's something I also thought about. Adding an option to explicitly select an driver seems like a good decision.
But I'd like to keep the fallback as default `"auto"` option for easier usage. If pulseaudio is installed it should definitely be used, as the default device is maybe only available in pulseaudio (e.g. bluetooth headphones). But having pulseaudio as default would break alsa only systems when no driver is set.

atheriel commented:
Yep, I'm happy with the fallback option being the default. That was what I'd expect if the user does not explicitly specify a `driver`.

atheriel commented:
@SWW13 Happy to merge this now. Did you want me to keep the history intact or can I squash the commits?

SWW13 commented:
I'd prefer squashing.

atheriel commented:
Thanks for all your work on this. :+1:
